src/fs.rs

```diff
@@ -0,0 +1,184 @@
+//! # ByteCache File System Operations
+//!
+//! This module handles file system interactions for the ByteCache system,
+//! providing interfaces for reading from and writing to cache storage on disk.
+//!
+//! It implements two primary access patterns:
+//!
+//! * **Read-only access**: Used for sideloaded content that should not be modified
+//! * **Read-write access**: Used for the persistent disk cache
+//!
+//! The module uses a marker type pattern to distinguish between these access modes
+//! at compile time, ensuring that operations are only performed when appropriate.
+
+use std::path::PathBuf;
+
+/// Marker type for read-only filesystem operations.
+///
+/// This is used for cache layers that should only read pre-existing data,
+/// such as the sideload cache.
+pub struct Read;
+
+/// Marker type for read-write filesystem operations with capacity limit.
+///
+/// This is used for cache layers that need to both read and write data,
+/// such as the disk cache. It includes a limit on the number of items
+/// to enforce cache size constraints.
+pub struct ReadWrite {
+    /// Maximum number of items to store in this cache
+    pub limit: usize,
+}
+
+/// File system cache representation.
+///
+/// This struct represents a directory-based cache on the filesystem,
+/// with operations determined by the generic type parameter T.
+/// T can be either `Read` (read-only) or `ReadWrite` (read-write).
+pub struct FsCache<T> {
+    /// Path to the directory containing cached items
+    pub path: PathBuf,
+    /// Type marker that determines available operations
+    pub kind: T,
```

Comment:
```suggestion
    path: PathBuf,
    kind: T,
```

How this type is implemented should be private.

Reply:
Should be fixed in next commit.